### ISSUE TITLE
Investigate high exclusive time in MSD.GetNode

### ISSUE THREAD

**Issue by rynowak** [aspnet/Mvc#5362]
This is sampling data from the LargeJsonApi app using the POST test scenario:
![image](https://cloud.githubusercontent.com/assets/1430011/19081435/1e030706-8a0f-11e6-91b7-5b32e98ab015.png)

The total time spent processing the request was 28s, and 3780ms of exclusive time (own time) was spent in `ModelStateDictionary.GetNode(...)`. This is 13.5% of the app's processing time. https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs#L509

You can see from the trace that `new StringSegment` and `ModelStateNode.GetNode` were not inlined, so I immediately suspect `string.IndexOfAny` which is implemented natively by the runtime. Our search space is generally constrained to dot-separate property names, which means the search text is small, and the number of delimiters we're looking for is 2. This seems like a good candidate for hand-optimizing.

**Comment by dougbu** [comment]
> Our search space is generally constrained to dot-separate property names

Need to handle look-ups for random variables, collection elements, and random mixes of supported expression types e.g. `@Html.TextBoxFor(m => variable1[variable2].Property.SubProperty)`.

**Comment by benaadams** [comment]
```
Inlines into 060000F9 ModelStateDictionary:GetNode(ref,bool):ref:this
  [0 IL=0056 TR=000058 06000018] [FAILED: unprofitable inline] StringSegment:.ctor(ref,int,int):this
  [0 IL=0065 TR=000066 060002EF] [FAILED: unprofitable inline] ModelStateNode:GetNode(struct,bool):ref:this
  [1 IL=0087 TR=000021 06001382] [profitable inline] String:IndexOfAny(ref,int):int:this
  [0 IL=0144 TR=000167 06000018] [FAILED: unprofitable inline] StringSegment:.ctor(ref,int,int):this
  [0 IL=0153 TR=000175 060002EF] [FAILED: unprofitable inline] ModelStateNode:GetNode(struct,bool):ref:this
  [2 IL=0163 TR=000129 060002E8] [below ALWAYS_INLINE size] ModelStateNode:get_Key():ref:this
  [3 IL=0172 TR=000139 060002E9] [below ALWAYS_INLINE size] ModelStateNode:set_Key(ref):this
```

**Comment by benaadams** [comment]
Made the `new StringSegment` inline for ya https://github.com/aspnet/Common/pull/159

**Comment by benaadams** [comment]
Also `string.IndexOfAny` initialises a probabilistic map before it begins searching

https://github.com/dotnet/coreclr/blob/master/src/classlibnative/bcltype/stringnative.cpp#L485-L537

If your `Delimiters` are fairly constant you could extract that out and use it multiple times; or even Vectorize the test...

**Comment by rynowak** [comment]
WOW thanks for digging into this. We're looking into relatively small strings (~6-100 characters) for 2 possible delimiters. I'm wondering how a naive implementation would compare. This method is called often, but the inputs are not complex.

**Comment by benaadams** [comment]
Have something for this... got very distracted on it O_o

**Comment by benaadams** [comment]
Resolved by merge of #5389
